File: common/src/main/java/org/sqlbuilder/common/DeSerialize.java
package org.sqlbuilder.common;

import java.io.*;

/**
 * Deserializer
 */
public class DeSerialize
{
	/**
	 * Safe cast
	 *
	 * @param object object
	 * @param <T>    cast type
	 * @return cast object
	 */
	@SuppressWarnings("unchecked")
	private static <T> T safeCast(Object object)
	{
		return (T) object;
	}

	/**
	 * Deserialize
	 *
	 * @param inFile serialization file
	 * @param <T>    type of result
	 * @return deserialized object
	 * @throws IOException            io exception
	 * @throws ClassNotFoundException class not found exception
	 */
	public static <T> T deserialize(final File inFile) throws IOException, ClassNotFoundException
	{
		return safeCast(deSerialize(inFile));
	}

	/**
	 * Deserialize object from file
	 *
	 * @param inFile input file
	 * @return object
	 * @throws IOException            io exception
	 * @throws ClassNotFoundException class not found exception
	 */
	public static Object deSerialize(final File inFile) throws IOException, ClassNotFoundException
	{
		try (InputStream is = new FileInputStream(inFile))
		{
			return deSerialize(is);
		}
	}

	/**
	 * Deserialize object from input stream
	 *
	 * @param is input stream
	 * @return object
	 * @throws IOException            io exception
	 * @throws ClassNotFoundException class not found exception
	 */
	public static Object deSerialize(final InputStream is) throws IOException, ClassNotFoundException
	{
		try (ObjectInputStream ois = new ObjectInputStream(is))
		{
			return ois.readObject();
		}
	}
}

File: common/src/main/java/org/sqlbuilder/common/DeSerialize.kt
package org.sqlbuilder.common

import java.io.*

/**
 * Deserializer
 */
object DeSerialize {

    /**
     * Safe cast
     *
     * @param obj object
     * @param <T> cast type
     * @return cast object
     */
    private fun <T> safeCast(obj: Any): T {
        @Suppress("UNCHECKED_CAST")
        return obj as T
    }

    /**
     * Deserialize
     *
     * @param inFile serialization file
     * @param <T>    type of result
     * @return deserialized object
     * @throws IOException            io exception
     * @throws ClassNotFoundException class not found exception
    </T> */
    @Throws(IOException::class, ClassNotFoundException::class)
    fun <T> deserialize(inFile: File): T {
        return safeCast<T>(deSerialize(inFile))
    }

    /**
     * Deserialize object from file
     *
     * @param inFile input file
     * @return object
     * @throws IOException            io exception
     * @throws ClassNotFoundException class not found exception
     */
    @Throws(IOException::class, ClassNotFoundException::class)
    fun deSerialize(inFile: File): Any {
        FileInputStream(inFile).use {
            return deSerialize(it)
        }
    }

    /**
     * Deserialize object from input stream
     *
     * @param input input stream
     * @return object
     * @throws IOException            io exception
     * @throws ClassNotFoundException class not found exception
     */
    @Throws(IOException::class, ClassNotFoundException::class)
    fun deSerialize(input: InputStream): Any {
        ObjectInputStream(input).use {
            return it.readObject()
        }
    }
}
